File: src/main/java/binarysearch/Hoppable.java
package binarysearch;

public class Hoppable {

    public boolean solve(int[] nums) {
        return recursive(0, nums);
    }

    private boolean recursive(int index, int[] nums) {
        if (nums.length - 1 <= index) {
            return true;
        }

        if (nums[index] == 0) {
            return false;
        }

        for (int i = 1; i <= nums[index]; i++) {
            if (recursive(index + i, nums)) {
                return true;
            } else {
                nums[index + i] = 0;
            }
        }

        return false;
    }

}

File: src/main/golang/binarysearch/Hoppable.go
package binarysearch

func Hoppable(nums []int) bool {
	return recursive(0, nums)
}

func recursive(index int, nums []int) bool {
	if index >= len(nums)-1 {
		return true
	}

	if nums[index] <= 0 {
		return false
	}

	for i := nums[index]; i > 0; i-- {
		if recursive(index+i, nums) {
			return true
		} else {
			nums[index+i] = 0
		}
	}
	return false
}
